tools/nl/validator/placetest.py

```diff
@@ -0,0 +1,264 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import csv
+from dataclasses import dataclass
+import json
+import os
+from typing import Dict, List, Set
+
+from absl import app
+from absl import flags
+import datacommons as dc
+import requests
+import utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('input_csv', 'data/bootstrap.csv', 'Input tokens CSV file')
+flags.DEFINE_string('names_csv', 'data/llm_output.csv',
+                    'Input SV descriptions CSV file')
+flags.DEFINE_string('checkpoint_dir', 'checkpoint/', 'Generated files')
+flags.DEFINE_string('run_name', 'foo',
+                    'Unique name of the test, for continuation, etc.')
+flags.DEFINE_bool('do_places_in', False, 'Generate places in?')
+
+URL = 'https://api.datacommons.org/v1/recognize/places?key=AIzaSyCTI4Xz-UW_G2Q2RfknhcfdAnTHq5X5XuI'
+POP_URL = 'https://api.datacommons.org/v1/bulk/observations/point?key=AIzaSyCTI4Xz-UW_G2Q2RfknhcfdAnTHq5X5XuI'
+
+OUT_HEADER = [
+    'Query', 'OrigPlace', 'WrongPlace', 'BogusPlace', 'WrongPlaceName',
+    'BogusPlaceName', 'Success', 'Exception', 'EmptyResult', 'SV'
+]
+
+CHECKPOINT_INTERVAL = 30
+
+
+@dataclass
+class Context:
+  rows: List[Dict]
+  counters: Dict
+  bootstrap: Dict
+  input_csv: str
+  names_csv: str
+  output_csv: str
+  counters_json: str
+  out_header: List[str]
+  bad_dcids: Set[str]
+  dcid_names: Dict
+  dcid_pop: Dict
+
+
+def init_result(q, sv, pl):
+  return {
+      'Query': q,
+      'SV': sv,
+      'OrigPlace': pl,
+      'WrongPlace': '',
+      'BogusPlace': '',
+      'Success': '',
+      'Exception': '',
+      'EmptyResult': '',
+      'WrongPlaceName': '',
+      'BogusPlaceName': '',
+  }
+
+
+def init_counters():
+  counters = {}
+  for c in [
+      'Total', 'Exception', 'EmptyResult', 'WrongPlace', 'BogusPlace', 'Success'
+  ]:
+    counters[c] = 0
+  return counters
+
+
+def compose_query(v):
+  # clean up some stuff
+  v = v.lower()
+  if v.startswith('population: '):
+    v = v.replace('population: ', 'number of people who are ')
+  v += ' in'
+  return ' '.join(v.split())
+
+
+def query(sv, pl, sv_name, pl_name, counters, bad_dcids):
```

Comment:
Two comments:

1. nit: could much of the work being done under each conditional be moved to a helper function? That might help with the readability since the function is fairly long.
2. ideally, it would be nice to have a function level comment. This isn't terribly hard to follow but would be nice to validate against.

Reply:
Added TODO for (1) and addressed (2).